src/main/scala/com/monsanto/arch/cloudformation/model/resource/ECS.scala

```diff
@@ -0,0 +1,240 @@
+                                                Memory: M = None,
+                                                MemoryReservation: R = None,
+                                                Name: String,
+                                                PortMappings: Option[Seq[PortMapping]] = None,
+                                                Privileged: Option[Boolean] = None,
+                                                ReadonlyRootFilesystem: Option[Boolean] = None,
+                                                Ulimits: Option[Seq[Ulimit]] = None,
+                                                User: Option[Token[String]] = None,
+                                                VolumesFrom: Option[Seq[VolumesFrom]] = None,
+                                                WorkingDirectory: Option[Token[String]] = None)(implicit ev1: MemoryRequirement[M, R]): ContainerDefinition = ContainerDefinition(Command, Cpu, DisableNetworking, DnsSearchDomains, DnsServers, DockerLabels, DockerSecurityOptions, EntryPoint, Environment, Essential, ExtraHosts, Hostname, Image, Links, LogConfiguration, Memory, MemoryReservation, Name, PortMappings, Privileged, ReadonlyRootFilesystem, Ulimits, User, VolumesFrom, WorkingDirectory, None)
+  implicit val format: JsonFormat[ContainerDefinition] = new RootJsonFormat[ContainerDefinition] {
+    override def write(cd: ContainerDefinition) = {
+      val obj = JsObject(
+        "Command" → cd.Command.toJson,
+        "Cpu" → cd.Cpu.toJson,
+        "DisableNetworking" → cd.DisableNetworking.toJson,
+        "DnsSearchDomains" → cd.DnsSearchDomains.toJson,
+        "DnsServers" → cd.DnsServers.toJson,
+        "DockerLabels" → cd.DockerLabels.toJson,
+        "DockerSecurityOptions" → cd.DockerSecurityOptions.toJson,
+        "EntryPoint" → cd.EntryPoint.toJson,
+        "Environment" → cd.Environment.toJson,
+        "Essential" → cd.Essential.toJson,
+        "ExtraHosts" → cd.ExtraHosts.toJson,
+        "Hostname" → cd.Hostname.toJson,
+        "Image" → cd.Image.toJson,
+        "Links" → cd.Links.toJson,
+        "LogConfiguration" → cd.LogConfiguration.toJson,
+        "Memory" → cd.Memory.toJson,
+        "MemoryReservation" → cd.MemoryReservation.toJson,
+        "Name" → cd.Name.toJson,
+        "PortMappings" → cd.PortMappings.toJson,
+        "Privileged" → cd.Privileged.toJson,
+        "ReadonlyRootFilesystem" → cd.ReadonlyRootFilesystem.toJson,
+        "Ulimits" → cd.Ulimits.toJson,
+        "User" → cd.User.toJson,
+        "VolumesFrom" → cd.VolumesFrom.toJson,
+        "WorkingDirectory" → cd.WorkingDirectory.toJson
+      )
+      obj.copy(fields = obj.fields.filter(_._2 != JsNull))
+    }
+
+    //noinspection NotImplementedCode
+    override def read(json: JsValue) = ???
+  }
+
+  @implicitNotFound("one or both of Memory and MemoryReservation must be defined")
+  sealed trait MemoryRequirement[M, R]
```

Comment:
This list of objects implementing `MemoryRequirement[M, R]` is kind of ugly, but I couldn't find another way to do this. Making `MemoryRequirement[M, R]` covariant (i.e. `MemoryRequirement[+M, +R]`) allows either `Option(42)` or `Some(42)` to be passed by calling code, but introduces ambiguity when both `Memory` and `MemoryReservation` are set. Without the covariance or this list of objects, only the `Some` form could be passed to `Memory` or `MemoryReservation`, and I didn't want to restrict callers like that.